### PR TITLE
[SPARK-28475][CORE] Add regex MetricFilter to GraphiteSink

### PR DESCRIPTION
**Pull request by nkarpov**
## What changes were proposed in this pull request?

Today all registered metric sources are reported to GraphiteSink with no filtering mechanism, although the codahale project does support it.

GraphiteReporter (ScheduledReporter) from the codahale project requires you implement and supply the MetricFilter interface (there is only a single implementation by default in the codahale project, MetricFilter.ALL).

Propose to add an additional regex config to match and filter metrics to the GraphiteSink

## How was this patch tested?

Included a GraphiteSinkSuite that tests:

1. Absence of regex filter (existing default behavior maintained)
2. Presence of `regex=<regexexpr>` correctly filters metric keys

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by dongjoon-hyun** [comment]
Thank you for your first contribution, @nkarpov . Could you elaborate more about the verification steps in your PR?

File: core/src/main/scala/org/apache/spark/metrics/sink/GraphiteSink.scala
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/metrics/sink/GraphiteSink.scala]: This empty line is one of the required Apache Spark coding styles.
```
[error] /home/jenkins/workspace/SparkPullRequestBuilder/core/src/main/scala/org/apache/spark/metrics/sink/GraphiteSink.scala:25:0: There should at least one a single empty line separating groups 3rdParty and spark.
```

You can run the code style checker by the following.
```
$ dev/scalastyle
```

**Comment by nkarpov** [comment]
Hi @dongjoon-hyun! 

Setup: start a graphite server & add host, port etc. in `metrics.properties`

For each scenario below step through w/ debugger to confirm MetricFilter is registered and corresponding metrics are posted to graphite server: 

1. Did not provide `driver.sink.graphite.regex` in `metrics.properties` -> successfully defaults to MetricFilter.ALL & all (driver) metrics reported

2. Did provide `driver.sink.graphite.regex=streaming` in `metrics.properties`, enable streaming metrics configuration (`spark.conf.set("spark.sql.streaming.metricsEnabled", true)`) & start a MemoryStream to console -> successfully registers anonymous MetricFilter in this PR and reports the 6 metrics that are expected to match regex expression `streaming`

I couldn't find an existing testing suite for the GraphiteSink so I verified manually as above. The change seemed minor enough for that to be OK but let me know if a more robust suite should be added.

Also, if we're happy with the naming convention here, I will add documentation in the following files as part of the PR

https://github.com/apache/spark/blob/master/docs/monitoring.md &
https://github.com/apache/spark/blob/master/conf/metrics.properties.template

**Comment by HyukjinKwon** [comment]
@nkarpov, please update this comment in the PR description under `How was this patch tested?`. 

Also, since there is no test, it would be better to describe, with commands, step by step from scratch. With explicit steps, we and our descendants later can detect regressions later other fixes come. Otherwise, every time we should parse your steps and come up with the proper commands.

Ideally it should be copy-and-pastable to reduce human efforts since there's no test case that detects regression automatically.

**Comment by nkarpov** [comment]
Thanks @HyukjinKwon - in that case it's worth to just add the tests. Added in latest commit.

**Comment by HyukjinKwon** [comment]
Thanks for adding a test.

**Comment by HyukjinKwon** [comment]
retest this please

File: core/src/test/scala/org/apache/spark/metrics/sink/GraphiteSinkSuite.scala
Comment (HeartSaVioR) [inline core/src/test/scala/org/apache/spark/metrics/sink/GraphiteSinkSuite.scala]: It would be ideal to have regex pattern on test case, as we support regex.

**Comment by HeartSaVioR** [comment]
Thanks for the contribution. I'm seeing the benefit on the change as I introduced similar thing (not for  Spark though), but it also ended up with advanced (complicated) supports: accepting multiple patterns, with mode either "whitelist" or "blacklist" (not both, of course). 

Hopefully I'm not seeing too many Dropwizard metrics in Spark so this might be good enough to start.

File: core/src/test/scala/org/apache/spark/metrics/sink/GraphiteSinkSuite.scala
Comment (nkarpov) [reply core/src/test/scala/org/apache/spark/metrics/sink/GraphiteSinkSuite.scala]: Done!

**Comment by nkarpov** [comment]
Thanks @HeartSaVioR - I've added a better regex example in the test. And you make a good point re: whitelist/blacklist - I think it'll be clear with an example in metrics.properties template? Should we rename the config?

Otherwise, @HeartSaVioR @HyukjinKwon @dongjoon-hyun, it's my first commit so please help me understand what it takes to be merged :) My only outstanding item ATM is to include an example in metrics.properties template file if we are good on naming convention.

**Comment by HeartSaVioR** [comment]
I'm also one of contributors trying to help reviewing. :) @HyukjinKwon and @dongjoon-hyun are able to review and approve the patch. So let's wait their next round of reviews.

**Comment by HyukjinKwon** [comment]
Can we make the tests passed? Also, you can cc people who appears in Git blame. cc @jerryshao

**Comment by nkarpov** [comment]
I've appended some documentation. I don't think the previous failures were related

**Review by HeartSaVioR — APPROVED** [review]
LGTM, but as I said earlier you need to get approvals from committers.

**Comment by jerryshao** [comment]
LGTM. Trigger the test again.

**Comment by jerryshao** [comment]
Jenkins, retest this please.

**Review by jerryshao — APPROVED** [review]
LGTM.

**Comment by jerryshao** [comment]
Thanks for the contribution, merging to master branch.

**Comment by dongjoon-hyun** [comment]
Thank you so much, @jerryshao ! :D

**Pull request closed without merging**
